crates/ironrdp-web/src/session.rs

```diff
@@ -492,18 +477,120 @@ impl Session {
-                            .map_err(|e| anyhow::Error::msg(format!("hide pointer callback failed: {e:?}")))?;
+                            .set_cursor_style_callback
+                            .call3(
+                                &self.set_cursor_style_callback_context,
+                                &JsValue::from_str("none"),
+                                &JsValue::UNDEFINED,
+                                &JsValue::UNDEFINED,
+                            )
+                            .map_err(|e| anyhow::Error::msg(format!("set cursor style callback failed: {e:?}")))?;
                     }
                     ActiveStageOutput::PointerPosition { .. } => {
-                        // Not applicable for web
+                        // Not applicable for web.
+                    }
+                    ActiveStageOutput::NativePointerUpdate(pointer) => {
+                        // Maximum allowed cursor size for browsers is 32x32, because bigger sizes
+                        // will cause the following issues:
+                        // - 128x128 and bigger cursors are not supported in browsers.
+                        // - 32x32 and bigger cursors will default to the system cursor if their
+                        //   sprite does not fit in the browser's viewport, introducing anabrupt
+                        //   cursor style changes when the cursor is moved to the edge of the
+                        //   browser window.
+                        //
+                        // Therefore, we need to scale the cursor sprite down to 32x32 if it is
+                        // bigger than that.
+                        const MAX_CURSOR_SIZE: u16 = 32;
+                        // INVARIANT: 0 < scale <= 1.0
+                        let scale = if pointer.width >= pointer.height && pointer.width > MAX_CURSOR_SIZE {
+                            Some(f64::from(MAX_CURSOR_SIZE) / f64::from(pointer.width))
+                        } else if pointer.height > MAX_CURSOR_SIZE {
+                            Some(f64::from(MAX_CURSOR_SIZE) / f64::from(pointer.height))
+                        } else {
+                            None
+                        };
+
+                        let (png_width, png_height, hot_spot_x, hot_spot_y, rgba_buffer) = if let Some(scale) = scale {
+                            // Per invariants: Following conversions are losslesss.
+                            let scaled_width = cast_f64_to_u16_lossy(f64::from(pointer.width) * scale);
+                            let scaled_height = cast_f64_to_u16_lossy(f64::from(pointer.height) * scale);
+                            let hotspot_x = cast_f64_to_u16_lossy(f64::from(pointer.hotspot_x) * scale);
+                            let hotspot_y = cast_f64_to_u16_lossy(f64::from(pointer.hotspot_y) * scale);
+
+                            // INVARIANT: scaled_width * scaled_height * 4 <= 32 * 32 * 4
```

Comment:
Note (style): here, we are [not defining an invariant, but using one](https://github.com/Devolutions/IronRDP/blob/08d165d71f70d4b819bcd689f339e1a4ce06ece5/STYLE.md#explain-non-obvious-assumptions-by-referencing-the-invariants).

```suggestion
                            // Per invariants: scaled_width * scaled_height * 4 <= 32 * 32 * 4 < usize::MAX
```

It’s also useful specify the `< usize::MAX` part here, because that’s the reason why there is no overflow (i.e.: 32 * 32 * 4 is smaller than `usize::MAX` on any architecture, which means we are not overflowing).

Reply:
My idea during writing of this was that `32 * 32 * 4` part is enough to explicitly state `< usize::MAX`

Reply:
You are right, indirectly we have `< usize::MAX`. However, ` < usize::MAX` is also the important part (the reason why no overflow can happen), so it’s useful to mention it, so the justification is "complete".